File: python/protein-translation/protein_translation.py
CODON_PROTEINS = {
    ("AUG"): "Methionine",
    ("UUU, UUC"): "Phenylalanine",
    ("UUA, UUG"): "Leucine",
    ("UCU, UCC, UCA, UCG"): "Serine",
    ("UAU, UAC"): "Tyrosine",
    ("UGU, UGC"): "Cysteine",
    ("UGG"): "Tryptophan",
    ("UAA, UAG, UGA"): "STOP"
}

def proteins(strand):
    codon_array = []
    for i in range(0, len(strand), 3):
        codon = strand[i:i + 3]

        for (codons, protein) in CODON_PROTEINS.items():
            if codon in codons:
                if protein == "STOP":
                    return codon_array
                codon_array.append(protein)

    return codon_array

File: ruby/protein-translation/protein_translation.rb
class InvalidCodonError < StandardError
end

class Translation
  CODON_PROTEINS = {
    ['AUG'] =>                      'Methionine',
    ['UUU', 'UUC'] =>               'Phenylalanine',
    ['UUA', 'UUG'] =>               'Leucine',
    ['UCU', 'UCC', 'UCA', 'UCG'] => 'Serine',
    ['UAU', 'UAC'] =>               'Tyrosine',
    ['UGU', 'UGC'] =>               'Cysteine',
    ['UGG'] =>                      'Tryptophan',
    ['UAA', 'UAG', 'UGA'] =>        'STOP'
  }

  def self.of_codon(codon)
    raise InvalidCodonError unless codon.scan(/[AUGC]/) == codon.chars

    CODON_PROTEINS.each { |codons, protein| return protein if codons.include?(codon) }
  end

  def self.of_rna(strand)
    proteins = strand.chars.each_slice(3).each_with_object([]) do |codon, proteins|
      protein = of_codon(codon.join)
      return proteins if protein == 'STOP'
      proteins << protein
    end
  end
end
